src/pcap/client.go

```diff
@@ -0,0 +1,240 @@
+	}
+
+	return nil
+}
+
+func handleStream(stream API_CaptureClient, packetWriter *pcapgo.Writer, copyWg *sync.WaitGroup, cancel CancelCauseFunc) {
+	for {
+		time.Sleep(PollDelay)
+
+		res, err := stream.Recv()
+		if errors.Is(err, io.EOF) {
+			zap.L().Info("clean stop, done") // TODO: fix logger
+			cancel(nil)
+			break
+		}
+		code := status.Code(err)
+		if code != codes.OK {
+			err = fmt.Errorf("receive non-OK code: %v: %w", code, err)
+			cancel(err)
+			break
+		}
+
+		switch p := res.Payload.(type) {
+		case *CaptureResponse_Message:
+			writeMessage(p.Message)
+		case *CaptureResponse_Packet:
+			writePacket(p.Packet, packetWriter)
+		}
+	}
+	copyWg.Done()
+}
+
+func writeMessage(message *Message) {
+	var logLevel zapcore.Level
+
+	switch message.Type {
+	case MessageType_UNKNOWN:
+		logLevel = zapcore.WarnLevel
+	case MessageType_INSTANCE_UNAVAILABLE:
+		logLevel = zapcore.WarnLevel
+	case MessageType_START_CAPTURE_FAILED:
+		logLevel = zapcore.WarnLevel
+	case MessageType_INVALID_REQUEST:
+		logLevel = zapcore.ErrorLevel
+	case MessageType_CONGESTED:
+		logLevel = zapcore.WarnLevel
+	case MessageType_LIMIT_REACHED:
+		logLevel = zapcore.WarnLevel
+	case MessageType_CAPTURE_STOPPED:
+		logLevel = zapcore.InfoLevel
+	case MessageType_CONNECTION_ERROR:
+		logLevel = zapcore.ErrorLevel
+	}
+	log.Infof("%s - %v", message.String(), logLevel.String()) // TODO: duplicate logging
+	zap.L().Log(logLevel, "received message", zap.String("message-type", message.Type.String()), zap.Any("message", message.Message))
+}
+
+func writePacket(packet *Packet, packetWriter *pcapgo.Writer) {
+	captureInfo := gopacket.CaptureInfo{
+		Timestamp:      packet.Timestamp.AsTime(),
+		CaptureLength:  len(packet.Data),
+		Length:         int(packet.Length),
+		InterfaceIndex: 0,
+		AncillaryData:  nil,
+	}
+	err := packetWriter.WritePacket(captureInfo, packet.Data)
+	if err != nil {
+		zap.L().Error("writing packet to file failed", zap.Error(err))
+	}
+	zap.L().Info("received packet", zap.Int("bytes", len(packet.Data)), zap.Time("capture-timestamp", packet.Timestamp.AsTime()))
+}
+
+// TODO: still needed?
+func (c *Client) logProgress(ctx context.Context) {
+	if c.packetFile == os.Stdout {
+		// writing progress information could interfere with packet output when both are written to stdout
+		log.Debug("writing captures to stdout, skipping write-progress logs")
+		return
+	}
+
+	ticker := time.Tick(5 * time.Second)
+	for {
+		select {
+		case <-ticker:
+			info, err := c.packetFile.Stat()
+			if err != nil {
+				log.Debug("pcap output file already closed: ", zap.Error(err))
+				return
+			}
+			log.Info(fmt.Sprintf("\033[2K\rWrote %s bytes to disk.", bytefmt.ByteSize(uint64(info.Size()))))
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func tryCreateOutputFile(outputFile string) (*os.File, error) {
+	if _, err := os.Stat(outputFile); os.IsNotExist(err) {
+		return os.Create(outputFile)
+	}
+	return nil, fmt.Errorf("output file %s already exists", outputFile)
```

Comment:
Why is this an error? Do we need some logic to force overwrite where this does not kick in?